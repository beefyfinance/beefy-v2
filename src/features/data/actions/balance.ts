import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { FetchAllBalancesResult } from '../apis/balance/balance-types';
import { getBalanceApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import { selectAllTokenWhereUserCouldHaveBalance } from '../selectors/balance';
import { selectBoostById, selectBoostsByChainId } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectTokenByAddress } from '../selectors/tokens';
import { selectAllGovVaultsByChainId } from '../selectors/vaults';
import { selectWalletAddress } from '../selectors/wallet';
import type { TokenEntity } from '../entities/token';
import type { VaultEntity } from '../entities/vault';
import { isGovVault } from '../entities/vault';
import { uniqueTokens } from '../../../helpers/tokens';

export interface FetchAllBalanceActionParams {
  chainId: ChainEntity['id'];
  walletAddress: string;
}

export interface FetchAllBalanceFulfilledPayload {
  chainId: ChainEntity['id'];
  walletAddress: string;
  data: FetchAllBalancesResult;
  // reducers need the state (balance)
  state: BeefyState;
}

export const fetchAllBalanceAction = createAsyncThunk<
  FetchAllBalanceFulfilledPayload,
  FetchAllBalanceActionParams,
  { state: BeefyState }
>('balance/fetchAllBalanceAction', async ({ chainId, walletAddress }, { getState }) => {
  const state = getState();

  const userAddress = walletAddress ?? selectWalletAddress(state);
  const chain = selectChainById(state, chainId);
  const api = await getBalanceApi(chain);

  const tokens = selectAllTokenWhereUserCouldHaveBalance(state, chainId).map(address =>
    selectTokenByAddress(state, chain.id, address)
  );
  // maybe have a way to retrieve those easily
  const boosts = selectBoostsByChainId(state, chainId).map(boostId =>
    selectBoostById(state, boostId)
  );
  const govVaults = selectAllGovVaultsByChainId(state, chain.id);

  const data = await api.fetchAllBalances(getState(), tokens, govVaults, boosts, userAddress);
  return {
    chainId,
    walletAddress: userAddress,
    data,
    state: getState(),
  };
});

export type FetchBalanceParams = {
  chainId: ChainEntity['id'];
  tokens?: TokenEntity[];
  vaults?: VaultEntity[];
};

export const fetchBalanceAction = createAsyncThunk<
  FetchAllBalanceFulfilledPayload,
  FetchBalanceParams,
  { state: BeefyState }
>(
  'balance/fetchBalanceAction',
  async ({ chainId, tokens: requestedTokens = [], vaults = [] }, { getState }) => {
    const state = getState();

    const walletAddress = selectWalletAddress(state);
    const chain = selectChainById(state, chainId);
    const api = await getBalanceApi(chain);

    const tokens = requestedTokens;
    const govVaults = [];
    const boosts = [];

    if (vaults.length) {
      for (const vault of vaults) {
        if (isGovVault(vault)) {
          govVaults.push(vault);
        } else {
          tokens.push(selectTokenByAddress(state, chain.id, vault.depositTokenAddress));
          tokens.push(selectTokenByAddress(state, chain.id, vault.earnedTokenAddress));
        }
      }
    }

    const data = await api.fetchAllBalances(
      getState(),
      uniqueTokens(tokens),
      govVaults,
      boosts,
      walletAddress
    );

    return {
      chainId,
      walletAddress,
      data,
      state: getState(),
    };
  }
);
