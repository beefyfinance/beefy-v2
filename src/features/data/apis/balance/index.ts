import { BalanceAPI } from './balance';
import { BalanceMcV2API } from './balance-multicallv2';

export { BalanceAPI, BalanceMcV2API };
