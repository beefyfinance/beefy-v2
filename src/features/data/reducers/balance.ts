import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import { byDecimals } from '../../../helpers/format';
import { fetchAllBalanceAction } from '../actions/balance';
import { BoostEntity } from '../entities/boost';
import { ChainEntity } from '../entities/chain';
import { TokenEntity } from '../entities/token';
import { VaultEntity } from '../entities/vault';
import { selectBoostById } from '../selectors/boosts';
import { selectTokenById } from '../selectors/tokens';
import {
  selectIsStandardVaultEarnTokenId,
  selectVaultByEarnTokenId,
  selectVaultById,
  selectVaultPricePerFullShare,
} from '../selectors/vaults';
import { accountHasChanged, walletHasDisconnected } from './wallet';

/**
 * State containing user balances state
 */
export interface BalanceState {
  depositedVaultIds: {
    govVaults: { chainId: ChainEntity['id']; vaultId: VaultEntity['id'] }[];
    standardVaults: { chainId: ChainEntity['id']; vaultId: VaultEntity['id'] }[];
  };
  // computed total balance accounting for boost
  // mostly useful for the vault list
  deposited: {
    [vaultId: VaultEntity['id']]: { balance: BigNumber; shares: BigNumber };
  };

  byChainId: {
    [chainId: ChainEntity['id']]: {
      byTokenId: {
        [tokenId: TokenEntity['id']]: {
          balance: BigNumber;
        };
      };
      byBoostId: {
        [boostId: BoostEntity['id']]: {
          balance: BigNumber;
          rewards: BigNumber;
        };
      };
      byGovVaultId: {
        [vaultId: VaultEntity['id']]: {
          balance: BigNumber;
          rewards: BigNumber;
        };
      };
    };
  };
}
export const initialBalanceState: BalanceState = {
  byChainId: {},
  deposited: {},
  depositedVaultIds: { govVaults: [], standardVaults: [] },
};

export const balanceSlice = createSlice({
  name: 'balance',
  initialState: initialBalanceState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(accountHasChanged, sliceState => {
      sliceState.byChainId = {};
      sliceState.depositedVaultIds = { govVaults: [], standardVaults: [] };
    });
    builder.addCase(walletHasDisconnected, sliceState => {
      sliceState.byChainId = {};
      sliceState.depositedVaultIds = { govVaults: [], standardVaults: [] };
    });

    builder.addCase(fetchAllBalanceAction.fulfilled, (sliceState, action) => {
      const state = action.payload.state;
      const chainId = action.payload.chainId;

      /**
       * Ingest token data
       */
      for (const tokenBalance of action.payload.data.tokens) {
        if (sliceState.byChainId[chainId] === undefined) {
          sliceState.byChainId[chainId] = { byTokenId: {}, byBoostId: {}, byGovVaultId: {} };
        }

        // only update data if necessary
        const stateForToken = sliceState.byChainId[chainId].byTokenId[tokenBalance.tokenId];
        if (
          // only if balance is non-zero
          !tokenBalance.amount.isZero() &&
          // and state isn't already there
          // if it's there, only if amount differ
          (stateForToken === undefined || !stateForToken.balance.isEqualTo(tokenBalance.amount))
        ) {
          sliceState.byChainId[chainId].byTokenId[tokenBalance.tokenId] = {
            balance: tokenBalance.amount,
          };

          // now, if this is a vault token, we want to mark this vault as deposited
          if (selectIsStandardVaultEarnTokenId(state, chainId, tokenBalance.tokenId)) {
            const vaultId = selectVaultByEarnTokenId(state, chainId, tokenBalance.tokenId);

            sliceState.depositedVaultIds.standardVaults.push({ chainId, vaultId });

            // add to total balance
            const vault = selectVaultById(state, vaultId);
            const oracleToken = selectTokenById(state, vault.chainId, vault.oracleId);
            const ppfs = selectVaultPricePerFullShare(state, vault.id);

            if (!sliceState.deposited[vaultId]) {
              sliceState.deposited[vaultId] = {
                balance: new BigNumber(0),
                shares: new BigNumber(0),
              };
            }
            const balanceSingle = byDecimals(
              new BigNumber(tokenBalance.amount).multipliedBy(byDecimals(ppfs)).toFixed(8),
              oracleToken.decimals
            );
            const vaultTotalState = sliceState.deposited[vaultId];
            vaultTotalState.balance = vaultTotalState.balance.plus(balanceSingle);
            vaultTotalState.shares = vaultTotalState.shares.plus(tokenBalance.amount);
          }
        }
      }

      /**
       * Ingest boost data
       */
      for (const boostBalance of action.payload.data.boosts) {
        if (sliceState.byChainId[chainId] === undefined) {
          sliceState.byChainId[chainId] = { byTokenId: {}, byBoostId: {}, byGovVaultId: {} };
        }

        // only update data if necessary
        const stateForBoost = sliceState.byChainId[chainId].byBoostId[boostBalance.boostId];
        if (
          stateForBoost === undefined ||
          !stateForBoost.balance.isEqualTo(boostBalance.balance) ||
          !stateForBoost.rewards.isEqualTo(boostBalance.rewards)
        ) {
          sliceState.byChainId[chainId].byBoostId[boostBalance.boostId] = {
            balance: boostBalance.balance,
            rewards: boostBalance.rewards,
          };
        }

        // add to total balance of the target vault
        const boost = selectBoostById(state, boostBalance.boostId);
        const vault = selectVaultById(state, boost.vaultId);
        const oracleToken = selectTokenById(state, vault.chainId, vault.oracleId);
        const ppfs = selectVaultPricePerFullShare(state, vault.id);
        if (!sliceState.deposited[vault.id]) {
          sliceState.deposited[vault.id] = {
            balance: new BigNumber(0),
            shares: new BigNumber(0),
          };
        }
        const balanceSingle = byDecimals(
          boostBalance.balance.multipliedBy(byDecimals(ppfs)),
          oracleToken.decimals
        );
        const vaultTotalState = sliceState.deposited[vault.id];
        vaultTotalState.balance = vaultTotalState.balance.plus(balanceSingle);
        vaultTotalState.shares = vaultTotalState.shares.plus(boostBalance.balance);
      }

      /**
       * Ingest gov vaults data
       */
      for (const vaultBalance of action.payload.data.govVaults) {
        const vaultId = vaultBalance.vaultId;
        if (sliceState.byChainId[chainId] === undefined) {
          sliceState.byChainId[chainId] = { byTokenId: {}, byBoostId: {}, byGovVaultId: {} };
        }

        // only update data if necessary
        const stateForVault = sliceState.byChainId[chainId].byGovVaultId[vaultId];
        if (
          // only if balance is non-zero
          !vaultBalance.rewards.isZero() &&
          // and state isn't already there
          // if it's there, only if amount differ
          (stateForVault === undefined || !stateForVault.rewards.isEqualTo(vaultBalance.rewards))
        ) {
          const vaultState = {
            rewards: vaultBalance.rewards,
            balance: vaultBalance.balance,
          };
          sliceState.byChainId[chainId].byGovVaultId[vaultId] = vaultState;
          sliceState.depositedVaultIds.govVaults.push({ chainId, vaultId });

          // add to total balance
          const vault = selectVaultById(state, vaultId);
          const oracleToken = selectTokenById(state, vault.chainId, vault.oracleId);

          if (!sliceState.deposited[vaultId]) {
            sliceState.deposited[vaultId] = {
              balance: new BigNumber(0),
              shares: new BigNumber(0),
            };
          }
          const balanceSingle = byDecimals(vaultBalance.balance, oracleToken.decimals);
          const vaultTotalState = sliceState.deposited[vaultId];
          vaultTotalState.balance = vaultTotalState.balance.plus(balanceSingle);
          vaultTotalState.shares = vaultTotalState.shares.plus(vaultBalance.balance);
        }
      }
    });
  },
});
