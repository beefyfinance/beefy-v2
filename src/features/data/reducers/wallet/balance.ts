import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import type { Draft } from 'immer';
import { uniq } from 'lodash-es';
import type { BeefyState } from '../../../../redux-types';
import type { FetchAllBalanceFulfilledPayload } from '../../actions/balance';
import { fetchAllBalanceAction, fetchBalanceAction } from '../../actions/balance';
import { initiateBoostForm } from '../../actions/boosts';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../../actions/tokens';
import type {
  BoostBalance,
  GovVaultPoolBalance,
  TokenBalance,
} from '../../apis/balance/balance-types';
import type { BoostEntity } from '../../entities/boost';
import type { ChainEntity } from '../../entities/chain';
import type { TokenEntity } from '../../entities/token';
import type { VaultEntity } from '../../entities/vault';
import { selectAllVaultBoostIds, selectBoostById } from '../../selectors/boosts';
import {
  selectIsStandardVaultEarnTokenAddress,
  selectStandardVaultByEarnTokenAddress,
  selectVaultById,
} from '../../selectors/vaults';
import { initiateMinterForm } from '../../actions/minters';
import { initiateBridgeForm } from '../../actions/bridge';
import { selectMinterById } from '../../selectors/minters';
import { BIG_ZERO } from '../../../../helpers/big-number';

/**
 * State containing user balances state
 */
export interface BalanceState {
  // we want to keep everything by address to be able to display the right
  // data even when the user quickly changes account
  byAddress: {
    [address: string]: {
      // quick access to all deposited vaults for this address
      // this can include gov, standard, or a boost's target vault
      depositedVaultIds: VaultEntity['id'][];

      /**
       * all balances below represent token amounts
       */
      tokenAmount: {
        /**
         * Token balance, used to know standard vault balance with earnToken (mooXyzToken)
         * and oracle balance, to display how much the user can put in a vault or boost
         */
        byChainId: {
          [chainId: ChainEntity['id']]: {
            byTokenAddress: {
              [tokenAddress: TokenEntity['address']]: {
                balance: BigNumber;
              };
            };
          };
        };

        /**
         * The boost balance to know how much we withdraw from the boost
         * and how much rewards we can claim
         */
        byBoostId: {
          [boostId: BoostEntity['id']]: {
            balance: BigNumber;
            rewards: BigNumber;
          };
        };

        /**
         * The gov vault token balance and pending rewards
         */
        byGovVaultId: {
          [vaultId: VaultEntity['id']]: {
            balance: BigNumber;
            rewards: BigNumber;
          };
        };
      };
    };
  };
}

export const initialBalanceState: BalanceState = {
  byAddress: {},
};

export const balanceSlice = createSlice({
  name: 'balance',
  initialState: initialBalanceState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(fetchAllBalanceAction.fulfilled, (sliceState, action) => {
      addBalancesToState(sliceState, action.payload);
    });

    builder.addCase(fetchBalanceAction.fulfilled, (sliceState, action) => {
      addBalancesToState(sliceState, action.payload);
    });

    builder.addCase(initiateBoostForm.fulfilled, (sliceState, action) => {
      const state = action.payload.state;
      if (!action.payload.walletAddress) {
        return;
      }
      const boost = selectBoostById(action.payload.state, action.payload.boostId);
      const vault = selectVaultById(action.payload.state, boost.vaultId);
      const walletAddress = action.payload.walletAddress.toLowerCase();

      const walletState = getWalletState(sliceState, walletAddress);
      const balance = action.payload.balance;
      addTokenBalanceToState(state, walletState, vault.chainId, balance.tokens);
      addGovVaultBalanceToState(walletState, balance.govVaults);
      addBoostBalanceToState(state, walletState, balance.boosts);
    });

    builder.addCase(initiateMinterForm.fulfilled, (sliceState, action) => {
      const state = action.payload.state;
      if (!action.payload.walletAddress) {
        return;
      }
      const minter = selectMinterById(action.payload.state, action.payload.minterId);
      const walletAddress = action.payload.walletAddress.toLowerCase();

      const walletState = getWalletState(sliceState, walletAddress);
      const balance = action.payload.balance;
      addTokenBalanceToState(state, walletState, minter.chainId, balance.tokens);
    });

    builder.addCase(initiateBridgeForm.fulfilled, (sliceState, action) => {
      const state = action.payload.state;
      if (!action.payload.walletAddress) {
        return;
      }

      const walletAddress = action.payload.walletAddress.toLocaleLowerCase();

      const walletState = getWalletState(sliceState, walletAddress);
      const balance = action.payload.balance;

      addTokenBalanceToState(state, walletState, action.payload.chainId, balance.tokens);
    });

    builder.addCase(
      reloadBalanceAndAllowanceAndGovRewardsAndBoostData.fulfilled,
      (sliceState, action) => {
        const state = action.payload.state;
        const chainId = action.payload.chainId;
        const walletAddress = action.payload.walletAddress.toLowerCase();

        const walletState = getWalletState(sliceState, walletAddress);
        const balance = action.payload.balance;
        addTokenBalanceToState(state, walletState, chainId, balance.tokens);
        addGovVaultBalanceToState(walletState, balance.govVaults);
        addBoostBalanceToState(state, walletState, balance.boosts);
      }
    );
  },
});

function getWalletState(sliceState: Draft<BalanceState>, walletAddress: string) {
  if (sliceState.byAddress[walletAddress] === undefined) {
    sliceState.byAddress[walletAddress] = {
      depositedVaultIds: [],
      tokenAmount: {
        byChainId: {},
        byBoostId: {},
        byGovVaultId: {},
      },
    };
  }

  return sliceState.byAddress[walletAddress];
}

function addBalancesToState(
  sliceState: Draft<BalanceState>,
  payload: FetchAllBalanceFulfilledPayload
) {
  const state = payload.state;
  const chainId = payload.chainId;
  const walletAddress = payload.walletAddress.toLowerCase();
  const walletState = getWalletState(sliceState, walletAddress);
  const balance = payload.data;

  addTokenBalanceToState(state, walletState, chainId, balance.tokens);
  addBoostBalanceToState(state, walletState, balance.boosts);
  addGovVaultBalanceToState(walletState, balance.govVaults);
}

function addTokenBalanceToState(
  state: BeefyState,
  walletState: Draft<BalanceState['byAddress']['0xABC']>,
  chainId: ChainEntity['id'],
  balances: TokenBalance[]
) {
  /**
   * Ingest token data
   */
  for (const tokenBalance of balances) {
    if (walletState.tokenAmount.byChainId[chainId] === undefined) {
      walletState.tokenAmount.byChainId[chainId] = { byTokenAddress: {} };
    }

    // only update data if necessary
    const stateForToken =
      walletState.tokenAmount.byChainId[chainId].byTokenAddress[
        tokenBalance.tokenAddress.toLowerCase()
      ];
    if (
      // state isn't already there if it's there, only if amount differ
      stateForToken === undefined ||
      !stateForToken.balance.isEqualTo(tokenBalance.amount)
    ) {
      walletState.tokenAmount.byChainId[chainId].byTokenAddress[
        tokenBalance.tokenAddress.toLowerCase()
      ] = {
        balance: tokenBalance.amount,
      };

      // if the token is the earnedToken of a vault
      // this means the user deposited in this vault
      if (selectIsStandardVaultEarnTokenAddress(state, chainId, tokenBalance.tokenAddress)) {
        const vaultId = selectStandardVaultByEarnTokenAddress(
          state,
          chainId,
          tokenBalance.tokenAddress
        );

        // to decide if we want to add or remove the vault we consider both the boost and vault deposit
        // I know we are adding carrots and oignons here but it's just to check if > 0
        let totalDepositOrRewards = tokenBalance.amount;
        for (const boostId of selectAllVaultBoostIds(state, vaultId)) {
          const boostDeposit = walletState.tokenAmount.byBoostId[boostId]?.balance || BIG_ZERO;
          const boostRewards = walletState.tokenAmount.byBoostId[boostId]?.rewards || BIG_ZERO;
          totalDepositOrRewards = totalDepositOrRewards.plus(boostDeposit).plus(boostRewards);
        }
        addOrRemoveFromDepositedList(walletState, totalDepositOrRewards, vaultId);
      }
    }
  }
}

function addGovVaultBalanceToState(
  walletState: Draft<BalanceState['byAddress']['0xABC']>,
  govVaultBalance: GovVaultPoolBalance[]
) {
  for (const vaultBalance of govVaultBalance) {
    const vaultId = vaultBalance.vaultId;

    // bug with old bifi gov pool
    if (vaultId === 'bifi-gov-eol') {
      vaultBalance.rewards = BIG_ZERO;
    }

    // only update data if necessary
    const stateForVault = walletState.tokenAmount.byGovVaultId[vaultId];
    if (
      // state isn't already there if it's there, only if amount differ
      stateForVault === undefined ||
      !stateForVault.rewards.isEqualTo(vaultBalance.rewards)
    ) {
      const vaultState = {
        rewards: vaultBalance.rewards,
        balance: vaultBalance.balance,
      };
      walletState.tokenAmount.byGovVaultId[vaultId] = vaultState;

      addOrRemoveFromDepositedList(walletState, vaultBalance.balance, vaultBalance.vaultId);
    }
  }
}

function addBoostBalanceToState(
  state: BeefyState,
  walletState: Draft<BalanceState['byAddress']['0xABC']>,
  boostBalances: BoostBalance[]
) {
  for (const boostBalance of boostBalances) {
    // only update data if necessary
    const stateForBoost = walletState.tokenAmount.byBoostId[boostBalance.boostId];
    if (
      stateForBoost === undefined ||
      !stateForBoost.balance.isEqualTo(boostBalance.balance) ||
      !stateForBoost.rewards.isEqualTo(boostBalance.rewards)
    ) {
      walletState.tokenAmount.byBoostId[boostBalance.boostId] = {
        balance: boostBalance.balance,
        rewards: boostBalance.rewards,
      };
    }
  }

  // once we added all boosts, find out if we have staked something in each vault
  const allVaultIds = uniq(
    boostBalances.map(boostBalance => {
      const boost = selectBoostById(state, boostBalance.boostId);
      return boost.vaultId;
    })
  );
  for (const vaultId of allVaultIds) {
    const vault = selectVaultById(state, vaultId);
    const vaultBalance =
      walletState.tokenAmount.byChainId[vault.chainId]?.byTokenAddress[
        vault.earnedTokenAddress.toLowerCase()
      ]?.balance || BIG_ZERO;
    let totalDepositOrRewards = new BigNumber(vaultBalance);
    for (const boostId of selectAllVaultBoostIds(state, vaultId)) {
      const boostDeposit = walletState.tokenAmount.byBoostId[boostId]?.balance || BIG_ZERO;
      const boostRewards = walletState.tokenAmount.byBoostId[boostId]?.rewards || BIG_ZERO;
      totalDepositOrRewards = totalDepositOrRewards.plus(boostDeposit).plus(boostRewards);
      addOrRemoveFromDepositedList(walletState, totalDepositOrRewards, vaultId);
    }
  }
}

function addOrRemoveFromDepositedList(
  walletState: Draft<BalanceState['byAddress']['0xABC']>,
  amount: BigNumber,
  vaultId: VaultEntity['id']
) {
  if (amount.isGreaterThan(0)) {
    if (!walletState.depositedVaultIds.includes(vaultId)) {
      walletState.depositedVaultIds.push(vaultId);
    }
  } else {
    if (walletState.depositedVaultIds.includes(vaultId)) {
      walletState.depositedVaultIds = walletState.depositedVaultIds.filter(vid => vid !== vaultId);
    }
  }
}
