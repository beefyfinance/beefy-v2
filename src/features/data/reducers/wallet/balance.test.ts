import BigNumber from 'bignumber.js';
import { fetchAllBalanceAction, FetchAllBalanceFulfilledPayload } from '../../actions/balance';
import {
  fetchAllContractDataByChainAction,
  FetchAllContractDataFulfilledPayload,
} from '../../actions/contract-data';
import { getBeefyTestingStore } from '../../utils/test-utils';
import { balanceSlice, initialBalanceState } from './balance';

describe('Balance slice tests', () => {
  it('should update state on fulfilled token balance', async () => {
    const store = await getBeefyTestingStore();
    // given we have some prior info
    const initPayload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        boosts: [],
        govVaults: [],
        standardVaults: [
          {
            id: 'banana-bnb-stars',
            pricePerFullShare: new BigNumber(5),
            balance: new BigNumber(123),
            strategy: 'test',
          },
        ],
      },
      state: store.getState(),
    };
    store.dispatch({ type: fetchAllContractDataByChainAction.fulfilled, payload: initPayload });

    const payload: FetchAllBalanceFulfilledPayload = {
      chainId: 'bsc',
      walletAddress: '0x00ABC',
      data: {
        boosts: [],
        govVaults: [],
        tokens: [
          { tokenId: 'banana-nfty-wbnb', amount: new BigNumber(10) },
          { tokenId: 'BIFI', amount: new BigNumber(10) },
          { tokenId: 'mooApeBNB-STARS', amount: new BigNumber(10) },
        ],
      },
      state: store.getState(),
    };
    const action = { type: fetchAllBalanceAction.fulfilled, payload: payload };
    const state = balanceSlice.reducer(store.getState().user.balance, action);
    expect(state).toMatchSnapshot();

    expect(state.byAddress['0x00abc']).toBeDefined();
    expect(state.byAddress['0x00abcdef']).toBeUndefined();

    // getting the same vaults don't update the state object
    const beforeReDispatch = state.byAddress['0x00abc'].tokenAmount.byChainId['bsc'].byTokenId;
    const newState = balanceSlice.reducer(state, action);
    const afterReDispatch = newState.byAddress['0x00abc'].tokenAmount.byChainId['bsc'].byTokenId;
    expect(beforeReDispatch).toBe(afterReDispatch);
  });

  it('should update state on fulfilled gov boost pools', async () => {
    const store = await getBeefyTestingStore();

    // given we have some prior info
    const payload: FetchAllBalanceFulfilledPayload = {
      chainId: 'bsc',
      walletAddress: '0x00ABC',
      data: {
        boosts: [],
        tokens: [],
        govVaults: [
          { vaultId: 'bifi-gov', rewards: new BigNumber(100), balance: new BigNumber(10) },
        ],
      },
      state: store.getState(),
    };
    const action = { type: fetchAllBalanceAction.fulfilled, payload: payload };
    const state = balanceSlice.reducer(initialBalanceState, action);
    expect(state).toMatchSnapshot();

    expect(state.byAddress['0x00abc']).toBeDefined();
    expect(state.byAddress['0x00abcdef']).toBeUndefined();

    // getting the same vaults don't update the state object
    const beforeReDispatch = state.byAddress['0x00abc'].tokenAmount.byGovVaultId;
    const newState = balanceSlice.reducer(state, action);
    const afterReDispatch = newState.byAddress['0x00abc'].tokenAmount.byGovVaultId;
    expect(beforeReDispatch).toBe(afterReDispatch);
  });

  it('should update state on fulfilled boost balance', async () => {
    const store = await getBeefyTestingStore();
    // given we have some prior info
    const initPayload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        boosts: [],
        govVaults: [],
        standardVaults: [
          {
            id: 'ellipsis-renbtc',
            pricePerFullShare: new BigNumber(5),
            balance: new BigNumber(123),
            strategy: 'test',
          },
          {
            id: 'belt-beltbnb',
            pricePerFullShare: new BigNumber(50),
            balance: new BigNumber(1230),
            strategy: 'test',
          },
        ],
      },
      state: store.getState(),
    };
    store.dispatch({ type: fetchAllContractDataByChainAction.fulfilled, payload: initPayload });

    const payload: FetchAllBalanceFulfilledPayload = {
      chainId: 'bsc',
      walletAddress: '0x00ABC',
      data: {
        govVaults: [],
        tokens: [],
        boosts: [
          {
            boostId: 'moo_ellipsis-renbtc-charge',
            balance: new BigNumber(100),
            rewards: new BigNumber(10),
          },
          {
            boostId: 'moo_beltbnb-czodiac',
            balance: new BigNumber(100),
            rewards: new BigNumber(10),
          },
        ],
      },
      state: store.getState(),
    };
    const action = { type: fetchAllBalanceAction.fulfilled, payload: payload };
    const state = balanceSlice.reducer(initialBalanceState, action);
    expect(state).toMatchSnapshot();

    expect(state.byAddress['0x00abc']).toBeDefined();
    expect(state.byAddress['0x00abcdef']).toBeUndefined();

    // getting the same vaults don't update the state object
    const beforeReDispatch = state.byAddress['0x00abc'].tokenAmount.byBoostId;
    const newState = balanceSlice.reducer(state, action);
    const afterReDispatch = newState.byAddress['0x00abc'].tokenAmount.byBoostId;
    expect(beforeReDispatch).toBe(afterReDispatch);
  });

  it('should consider a vault as "staked" when boost balance is zero', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const payload: FetchAllBalanceFulfilledPayload = {
      chainId: 'bsc',
      state: store.getState(),
      walletAddress: '0x000000000000',
      data: {
        // we have staked in the vault
        tokens: [{ tokenId: 'mooApeBANANA-BUSD', amount: new BigNumber(1) }],
        govVaults: [],
        boosts: [
          {
            boostId: 'moo_banana-banana-busd-bitcrush',
            balance: new BigNumber(0), // but boost is empty
            rewards: new BigNumber(0),
          },
        ],
      },
    };

    // we should expect the vault to be included in the deposited list
    const action = { type: fetchAllBalanceAction.fulfilled, payload: payload };
    const newState = balanceSlice.reducer(initialBalanceState, action);
    expect(
      newState.byAddress['0x000000000000'].depositedVaultIds.includes('banana-banana-busd')
    ).toBeTruthy();
  });

  it('should consider a vault as "staked" when boost balance is zero and boost rewards is non-zero', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const payload: FetchAllBalanceFulfilledPayload = {
      chainId: 'bsc',
      state: store.getState(),
      walletAddress: '0x000000000000',
      data: {
        // we have NOT staked in the vault
        tokens: [{ tokenId: 'mooApeBANANA-BUSD', amount: new BigNumber(0) }],
        govVaults: [],
        boosts: [
          {
            boostId: 'moo_banana-banana-busd-bitcrush',
            balance: new BigNumber(0), // but boost is empty
            rewards: new BigNumber(1), // BUT we have some rewards
          },
        ],
      },
    };

    // we should expect the vault to be included in the deposited list
    const action = { type: fetchAllBalanceAction.fulfilled, payload: payload };
    const newState = balanceSlice.reducer(initialBalanceState, action);
    expect(
      newState.byAddress['0x000000000000'].depositedVaultIds.includes('banana-banana-busd')
    ).toBeTruthy();
  });

  it('should consider a vault as "staked" when the first boost balance is non-zero but last boost balance is zero', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const payload: FetchAllBalanceFulfilledPayload = {
      chainId: 'bsc',
      state: store.getState(),
      walletAddress: '0x000000000000',
      data: {
        // we have nothing in the vault
        tokens: [{ tokenId: 'mooApeBNB-STARS', amount: new BigNumber(0) }],
        govVaults: [],
        boosts: [
          {
            boostId: 'moo_banana-bnb-stars-mogul2',
            balance: new BigNumber(1), // this boost is not empty
            rewards: new BigNumber(1), // this boost is not empty
          },
          {
            boostId: 'moo_banana-bnb-stars-mogul',
            balance: new BigNumber(0), // but boost is empty
            rewards: new BigNumber(0),
          },
        ],
      },
    };

    // we should expect the vault to be included in the deposited list
    const action = { type: fetchAllBalanceAction.fulfilled, payload: payload };
    const newState = balanceSlice.reducer(initialBalanceState, action);
    expect(
      newState.byAddress['0x000000000000'].depositedVaultIds.includes('banana-bnb-stars')
    ).toBeTruthy();
  });
});
