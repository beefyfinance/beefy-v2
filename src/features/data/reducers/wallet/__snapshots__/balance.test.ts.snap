// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balance slice tests should update state on fulfilled boost balance 1`] = `
Object {
  "byAddress": Object {
    "0x00abc": Object {
      "depositedVaultIds": Array [
        "ellipsis-renbtc",
        "belt-beltbnb",
      ],
      "eligibleVaultIds": Array [],
      "tokenAmount": Object {
        "byBoostId": Object {
          "moo_beltbnb-czodiac": Object {
            "balance": "100",
            "rewards": "10",
          },
          "moo_ellipsis-renbtc-charge": Object {
            "balance": "100",
            "rewards": "10",
          },
        },
        "byChainId": Object {},
        "byGovVaultId": Object {},
      },
    },
  },
}
`;

exports[`Balance slice tests should update state on fulfilled gov boost pools 1`] = `
Object {
  "byAddress": Object {
    "0x00abc": Object {
      "depositedVaultIds": Array [
        "bifi-gov",
      ],
      "eligibleVaultIds": Array [],
      "tokenAmount": Object {
        "byBoostId": Object {},
        "byChainId": Object {},
        "byGovVaultId": Object {
          "bifi-gov": Object {
            "balance": "10",
            "rewards": "100",
          },
        },
      },
    },
  },
}
`;

exports[`Balance slice tests should update state on fulfilled token balance 1`] = `
Object {
  "byAddress": Object {
    "0x00abc": Object {
      "depositedVaultIds": Array [
        "banana-bnb-stars",
      ],
      "eligibleVaultIds": Array [
        "banana-nfty-wbnb",
        "bifi-maxi",
        "bifi-gov",
      ],
      "tokenAmount": Object {
        "byBoostId": Object {},
        "byChainId": Object {
          "bsc": Object {
            "byTokenId": Object {
              "BIFI": Object {
                "balance": "10",
              },
              "banana-nfty-wbnb": Object {
                "balance": "10",
              },
              "mooApeBNB-STARS": Object {
                "balance": "10",
              },
            },
          },
        },
        "byGovVaultId": Object {},
      },
    },
  },
}
`;
